### PR TITLE
Add ifunc support for Windows on AArch64.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
On Windows there is no platform support for ifunc but we could lower them to global function pointers.
This also enables FMV for Windows with Clang and Compiler-rt.

Depends on #111961

### CODE REVIEW

File: clang/test/CodeGen/ifunc-win.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
+// RUN: %clang_cc1 -triple aarch64-pc-windows-msvc -emit-llvm -o - %s | FileCheck %s
```
Comment (labrinea) [inline clang/test/CodeGen/ifunc-win.c:1]: Is this test file specific to AArch64? If so is it worth adding an 'aarch64-' prefix to it?

File: clang/include/clang/Basic/TargetInfo.h [hunk 2]
```diff
@@ -1505,6 +1505,10 @@ class TargetInfo : public TransferrableTargetInfo,
   bool supportsIFunc() const {
     if (getTriple().isOSBinFormatMachO())
       return true;
+    if (getTriple().isOSWindows() && getTriple().isAArch64())
+      return true;
+    if (getTriple().getArch() == llvm::Triple::ArchType::avr)
```
Comment (labrinea) [inline clang/include/clang/Basic/TargetInfo.h:1510]: Is this an unrelated change?

Comment (mstorsjo) [reply clang/include/clang/Basic/TargetInfo.h:1510]: I noted the same, but it seems it's intentional.

One of the existing tests use ifuncs for an `avr-unknown-unknown` target (which is OS-less). Previously, ifunc was accepted for any `TargetELFOrMachO`, which I guess was true for `avr-unknown-unknown`. But now it's only accepted for targets that are accepted by the `supportsIFunc()` method, which only accept a few specific OSes on ELF.

On the other hand, it doesn't seem consistent, that for baremetal, ifunc is supported depending on the architecture - but this is needed to keep the existing tests passing at least.

File: clang/test/CodeGen/ifunc-win.c (hunk 1 above)
Comment (DanielKristofKiss) [reply clang/test/CodeGen/ifunc-win.c:1]: So far, but other targets could be added here later without too much change to the test.

File: clang/include/clang/Basic/TargetInfo.h (hunk 2 above)
Comment (DanielKristofKiss) [reply clang/include/clang/Basic/TargetInfo.h:1510]: As @mstorsjo said above. 
Attribute support and the codegen now uses the same condition to check ifunc support and this caused test failures.
(Adding @benshi001 for AVR)

File: clang/include/clang/Basic/AttrDocs.td [hunk 3]
```diff
@@ -5796,12 +5796,18 @@ declared entity. The entity must not have weak linkage; for example, in C++,
...
 
-Not all targets support this attribute. ELF target support depends on both the
-linker and runtime linker, and is available in at least lld 4.0 and later,
-binutils 2.20.1 and later, glibc v2.11.1 and later, and FreeBSD 9.1 and later.
-Mach-O targets support it, but with slightly different semantics: the resolver
-is run at first call, instead of at load time by the runtime linker. Targets
-other than ELF and Mach-O currently do not support this attribute.
+Not all targets support this attribute:
+
+- ELF target support depends on both the linker and runtime linker, and is
```
Comment (jroelofs) [inline clang/include/clang/Basic/AttrDocs.td:5801]: Pending the outcome of the other thread, you should mention baremetal avr here.

Comment (DanielKristofKiss) [reply clang/include/clang/Basic/AttrDocs.td:5987]: added, due to it works in the last few releases so deserves an entry.
https://godbolt.org/z/K5dY7qbjj

**Comment by DanielKristofKiss** [comment]
ping

**Comment by labrinea** [comment]
I'll let others approve since I am not a windows expert, but it looks okay from my end at least.

File: clang/include/clang/Basic/AttrDocs.td (hunk 3 above)
Comment (MaskRay) [reply clang/include/clang/Basic/AttrDocs.td:5987]: I think we can drop the mention of lld here. lld 4.0 is not quite usable for a lot of issues/missing features.

**Review by MaskRay — APPROVED** [review]

**Review by jroelofs — APPROVED** [review]

**Comment by DanielKristofKiss** [comment]
test depends on the aarch64 backend in llvm.
// REQUIRES: aarch64-registered-target to be add to the test.

https://github.com/llvm/llvm-project/pull/116244
